### ISSUE TITLE
useJsxKeyInIterable false alarm in Map declaration

### ISSUE THREAD

**Issue by devuxer** [biomejs/biome#8572]
### Environment information

```bash
CLI:
  Version:                      2.3.10
  Color support:                true

Platform:
  CPU Architecture:             x86_64
  OS:                           windows

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  BIOME_THREADS:                unset
  NO_COLOR:                     unset
  TERM:                         unset
  JS_RUNTIME_VERSION:           v24.11.1
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         npm/11.6.1

Biome Configuration:
  Status:                       Loaded successfully
  Path:                         biome.json
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  false
  HTML full support enabled:    unset

Linter:
  JavaScript enabled:           unset
  JSON enabled:                 unset
  CSS enabled:                  unset
  GraphQL enabled:              unset
  Recommended:                  false
  Enabled rules:
    complexity/noAdjacentSpacesInRegex
    complexity/noExtraBooleanCast
    complexity/noUselessCatch
    complexity/noUselessTypeConstraint
    correctness/noChildrenProp
    correctness/noConstAssign
    correctness/noConstantCondition
    correctness/noEmptyCharacterClassInRegex
    correctness/noEmptyPattern
    correctness/noGlobalObjectCalls
    correctness/noInvalidBuiltinInstantiation
    correctness/noInvalidConstructorSuper
    correctness/noNonoctalDecimalEscape
    correctness/noPrecisionLoss
    correctness/noSelfAssign
    correctness/noSetterReturn
    correctness/noSwitchDeclarations
    correctness/noUndeclaredVariables
    correctness/noUnreachable
    correctness/noUnreachableSuper
    correctness/noUnsafeFinally
    correctness/noUnsafeOptionalChaining
    correctness/noUnusedLabels
    correctness/noUnusedPrivateClassMembers
    correctness/noUnusedVariables
    correctness/useExhaustiveDependencies
    correctness/useHookAtTopLevel
    correctness/useIsNan
    correctness/useJsxKeyInIterable
    correctness/useValidForDirection
    correctness/useValidTypeof
    correctness/useYield
    security/noBlankTarget
    security/noDangerouslySetInnerHtmlWithChildren
    style/noNamespace
    style/useAsConstAssertion
    style/useImportType
    suspicious/noAsyncPromiseExecutor
    suspicious/noCatchAssign
    suspicious/noClassAssign
    suspicious/noCommentText
    suspicious/noCompareNegZero
    suspicious/noConsole
    suspicious/noControlCharactersInRegex
    suspicious/noDebugger
    suspicious/noDuplicateCase
    suspicious/noDuplicateClassMembers
    suspicious/noDuplicateJsxProps
    suspicious/noDuplicateObjectKeys
    suspicious/noDuplicateParameters
    suspicious/noEmptyBlockStatements
    suspicious/noExtraNonNullAssertion
    suspicious/noFallthroughSwitchClause
    suspicious/noFunctionAssign
    suspicious/noGlobalAssign
    suspicious/noImportAssign
    suspicious/noMisleadingCharacterClass
    suspicious/noMisleadingInstantiator
    suspicious/noPrototypeBuiltins
    suspicious/noRedeclare
    suspicious/noShadowRestrictedNames
    suspicious/noSparseArray
    suspicious/noUnsafeDeclarationMerging
    suspicious/noUnsafeNegation
    suspicious/noWith
    suspicious/useAwait
    suspicious/useGetterReturn
    suspicious/useNamespaceKeyword

Workspace:
  Open Documents:               0
```

### Rule name

useJsxKeyInIterable

### Playground link

https://biomejs.dev/playground/?lintRules=useJsxKeyInIterable&code=CgBpAG0AcABvAHIAdAAgAHQAeQBwAGUAIAB7ACAAUgBlAGEAYwB0AE4AbwBkAGUAIAB9ACAAZgByAG8AbQAgACIAcgBlAGEAYwB0ACIAOwAKAHQAeQBwAGUAIABTAHQAYQB0AHUAcwAgAD0AIAAiAE4AbwBuAGUAIgAgAHwAIAAiAFAAYQByAHQAaQBhAGwAIgAgAHwAIAAiAEYAdQBsAGwAIgA7AAoACgBjAG8AbgBzAHQAIABzAHQAYQB0AHUAcwBUAG8ASQBjAG8AbgBNAGEAcAAgAD0AIABuAGUAdwAgAE0AYQBwADwAUwB0AGEAdAB1AHMALAAgAFIAZQBhAGMAdABOAG8AZABlAD4AKABbAAoAIAAgACAAIABbACIATgBvAG4AZQAiACwAIAA8AFQAcgBpAGEAbgBnAGwAZQBEAG8AdwBuAEkAYwBvAG4AIABzAHgAPQB7AHsAIABmAG8AbgB0AFMAaQB6AGUAOgAgACIAMQAuADIANQByAGUAbQAiACwAIABjAG8AbABvAHIAOgAgACIAZQByAHIAbwByAC4AbQBhAGkAbgAiACAAfQB9ACAALwA%2BAF0ALAAKACAAIAAgACAAWwAiAFAAYQByAHQAaQBhAGwAIgAsACAAPABUAHIAaQBhAG4AZwBsAGUARABvAHcAbgBJAGMAbwBuACAAcwB4AD0AewB7ACAAZgBvAG4AdABTAGkAegBlADoAIAAiADEALgAyADUAcgBlAG0AIgAsACAAYwBvAGwAbwByADoAIAAiAGUAcgByAG8AcgAuAG0AYQBpAG4AIgAgAH0AfQAgAC8APgBdACwACgAgACAAIAAgAFsAIgBGAHUAbABsACIALAAgADwAVAByAGkAYQBuAGcAbABlAEQAbwB3AG4ASQBjAG8AbgAgAHMAeAA9AHsAewAgAGYAbwBuAHQAUwBpAHoAZQA6ACAAIgAxAC4AMgA1AHIAZQBtACIALAAgAGMAbwBsAG8AcgA6ACAAIgBlAHIAcgBvAHIALgBtAGEAaQBuACIAIAB9AH0AIAAvAD4AXQAsAAoAXQApADsA&ruleDomains.react=recommended

### Expected result

It should not throw an error.

Linter seems to be confusing a Map entry tuple with an array.

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by creonico** [comment]
Hello! I’m interested in working on this issue as my first open-source contribution. It looks like a good fit for me.

Please let me know if I can take it. Thanks!

**Comment by Pranav2612000** [comment]
@creonico ! Are you still working on it? Or can I give it a shot?

**Comment by ematipico** [comment]
Go ahead @Pranav2612000

**Comment by Pranav2612000** [comment]
Great. I'm trying this out. 
When evaluating this rule, in `correctness/use_jsx_key_in_iterable` since we only get the actual array, there's no way to know if this is an actual array ( for which the rule must run ) or a part of the Map constructor arg ( for which the rule is to be ignored ).  I was thinking of updating the parsing grammar in `js.ungram` to add a new type for Map constructors and then ignore them for this rule. Do you think this is the right approach? Or do you have a better idea @ematipico ?

**Comment by siketyan** [comment]
@Pranav2612000 You can get the parent node of the array using `.parent()` and check if it is Map constructor

**Comment by Pranav2612000** [comment]
Something like this?
```
fn handle_collections(
    node: &JsArrayExpression,
    model: &SemanticModel,
    options: &UseJsxKeyInIterableOptions,
) -> Vec<TextRange> {
    let should_ignore_check = node
        .parent::<JsArrayElementList>()
        .and_then(|n| n.parent::<JsArrayExpression>())
        .and_then(|n| n.parent::<JsCallArgumentList>())
        .and_then(|n| n.parent::<JsCallArguments>())
        .and_then(|n| n.parent::<JsNewExpression>())
        .and_then(|n| n.callee().ok())
        .and_then(|c| c.get_callee_member_name())
        .and_then(|t| Some(t.token_text_trimmed() == "Map"))
        .unwrap_or(false);
    if should_ignore_check {
        return Vec::new();
    }
    // ... rest of the function
```

Don't like how the long `and_then` chain looks

**Comment by siketyan** [comment]
@Pranav2612000 ~Maybe these two lines are unnecessary given that the `node` is `JsArrayExpression`?~ Nvm, it's correct as  Map constructor requires an array of tuples

```rust
node
    .parent::<JsArrayElementList>()
    .and_then(|n| n.parent::<JsArrayExpression>())
```

**Comment by Pranav2612000** [comment]
> [@Pranav2612000](https://github.com/Pranav2612000) ~Maybe these two lines are unnecessary given that the `node` is `JsArrayExpression`?~ Nvm, it's correct as Map constructor requires an array of tuples
> 
> node
>     .parent::<JsArrayElementList>()
>     .and_then(|n| n.parent::<JsArrayExpression>())

Yup. Opening up a PR with this and some tests. Let's improve on it as part of the review.

**Comment by ematipico** [comment]
@dyc3 just so I understand, why is this a bug? 

Currently the rule flags components are declared inside an array. Should it be the same for maps and sets?

**Comment by Pranav2612000** [comment]
I think the constructor for Map and Sets is slightly different. [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/Map) expects an array of 2 elements, the first of which is the key and the second is the value. 

With Map, its sort of like you decide which one of the possible jsx elements is to be rendered based on some key ( different JSXs based on whether the key is "None", "Partial" or "Full" in the case of [this](https://biomejs.dev/playground/?lintRules=useJsxKeyInIterable&code=CgBpAG0AcABvAHIAdAAgAHQAeQBwAGUAIAB7ACAAUgBlAGEAYwB0AE4AbwBkAGUAIAB9ACAAZgByAG8AbQAgACIAcgBlAGEAYwB0ACIAOwAKAHQAeQBwAGUAIABTAHQAYQB0AHUAcwAgAD0AIAAiAE4AbwBuAGUAIgAgAHwAIAAiAFAAYQByAHQAaQBhAGwAIgAgAHwAIAAiAEYAdQBsAGwAIgA7AAoACgBjAG8AbgBzAHQAIABzAHQAYQB0AHUAcwBUAG8ASQBjAG8AbgBNAGEAcAAgAD0AIABuAGUAdwAgAE0AYQBwADwAUwB0AGEAdAB1AHMALAAgAFIAZQBhAGMAdABOAG8AZABlAD4AKABbAAoAIAAgACAAIABbACIATgBvAG4AZQAiACwAIAA8AFQAcgBpAGEAbgBnAGwAZQBEAG8AdwBuAEkAYwBvAG4AIABzAHgAPQB7AHsAIABmAG8AbgB0AFMAaQB6AGUAOgAgACIAMQAuADIANQByAGUAbQAiACwAIABjAG8AbABvAHIAOgAgACIAZQByAHIAbwByAC4AbQBhAGkAbgAiACAAfQB9ACAALwA%2BAF0ALAAKACAAIAAgACAAWwAiAFAAYQByAHQAaQBhAGwAIgAsACAAPABUAHIAaQBhAG4AZwBsAGUARABvAHcAbgBJAGMAbwBuACAAcwB4AD0AewB7ACAAZgBvAG4AdABTAGkAegBlADoAIAAiADEALgAyADUAcgBlAG0AIgAsACAAYwBvAGwAbwByADoAIAAiAGUAcgByAG8AcgAuAG0AYQBpAG4AIgAgAH0AfQAgAC8APgBdACwACgAgACAAIAAgAFsAIgBGAHUAbABsACIALAAgADwAVAByAGkAYQBuAGcAbABlAEQAbwB3AG4ASQBjAG8AbgAgAHMAeAA9AHsAewAgAGYAbwBuAHQAUwBpAHoAZQA6ACAAIgAxAC4AMgA1AHIAZQBtACIALAAgAGMAbwBsAG8AcgA6ACAAIgBlAHIAcgBvAHIALgBtAGEAaQBuACIAIAB9AH0AIAAvAD4AXQAsAAoAXQApADsA&ruleDomains.react=recommended) issue )


Another way to look at it is,
the following parses correctly
```
const x = {
	"None": <TriangleDownIcon sx={{ fontSize: "1.25rem", color: "error.main" }} />
}
```
but the following will fail
```
const x = new Map([
    ["None", <TriangleDownIcon sx={{ fontSize: "1.25rem", color: "error.main" }} />],
]);
```
even though both of them are similar.

**Comment by ematipico** [comment]
It doesn't matter how you construct an array or a map. What matters is how you render the elements. While the rule doesn't check if the array/iterator is actually used, it should check for every possible types that can be iterated.

`Map` and `Set` can be iterated, and they can be transformed into array too and be used with `.map` inside JSX.

**Comment by Pranav2612000** [comment]
> It doesn't matter how you construct an array or a map. What matters is how you render the elements. While the rule doesn't check if the array/iterator is actually used, it should check for every possible types that can be iterated.
> 
> `Map` and `Set` can be iterated, and they can be transformed into array too and be used with `.map` inside JSX.

I didn't completely understand what you're saying 😅 . Are you suggesting that `Set` should also have  similar logic as `Map` and we ignore this rule in their constructors? 
Or do you think we should always respect `useJsxKeyInIterable` rule and the arrays in constructors should always have a `key` property? If it's the latter, how do you explain the difference in behaviour when defining a simple object vs defining it through Map constructor?

 BTW, both approaches make sense to me. I'm sure you know which one is the best.

**Comment by dyc3** [comment]
Tbh, I think I was thinking of a different rule when I marked this as a bug, but the closer I look at it I'm still convinced its a bug.

I went and checked if the source rule flags this sample, which it does. The source rule's unit tests don't mention `Map` or `Set` at all https://github.com/jsx-eslint/eslint-plugin-react/blob/master/tests/lib/rules/jsx-key.js

However, I also went and tried it in a react playground:
```tsx
import React, { useState } from 'react'

const statusToIconMap = new Map([
	["None", <span>None</span>],
	["Partial", <span>Partial</span>],
	["Full", <span>Full</span>],
]);

export default function App() {
  const [status, setStatus] = useState("None");
  const [status2, setStatus2] = useState("None");
  return <div>
    <div>{statusToIconMap.get(status)}</div>
    <div>{statusToIconMap.get(status2)}</div>
    <button onClick={() => setStatus("None")}>set none</button>
    <button onClick={() => setStatus("Partial")}>set partial</button>
    <button onClick={() => setStatus("Full")}>set full</button>
    <button onClick={() => setStatus2("None")}>set none 2</button>
    <button onClick={() => setStatus2("Partial")}>set partial 2</button>
    <button onClick={() => setStatus2("Full")}>set full 2</button>
  </div>
}
```

This pattern works fine as long as the values in the Map are static and stateless. Therefore, I think the rule shouldn't flag this sample.

**Comment by ematipico** [comment]
@dyc3 that's not the right use case I am talking about. You should iterate the map, and then you iterate it, you must youse a key.

```jsx

const statusToIconMap = new Map([
	["None", <span>None</span>],
	["Partial", <span>Partial</span>],
	["Full", <span>Full</span>],
]);

const iconMap = Array.from(statusToIconMap.entries());

function App() {
	return iconMap.map(([_key, Cmp ]) => <Cmp />)
}
```

In this case, we would encounter the error where `key` isn't specified. 

> are static and stateless

The rule flags the definition of arrays, and they are static.

**Comment by dyc3** [comment]
Yeah I agree. the rule correctly flags the `return iconMap.map` but incorrectly flags the array in the `new Map`.

So by extension I guess it would also make sense to not flag anything that is outside the `.map()` call.

**Comment by dm-pse** [comment]
What about this?

```ts
const keyValuePairs = [
	["None", <span>None</span>],
	["Partial", <span>Partial</span>],
	["Full", <span>Full</span>],
];
const mapped = new Map(keyValuePairs);
```

This still is a case where `key` would not make sense, yet it wouldn't be declared inside a `new Map(...)`.

The only time a key should be needed is when an array actually gets rendered inside a component.

That said, I don't know if it's possible to trace that, so detecting based on whether a tuple is declared inside a Map constructor may be the best option.

EDIT:

I guess one other idea that comes to mind if to detect based on whether an array is a two-item tuple inside a parent array. If that is detectable, that might give the fewest false alarms without missing true violations.

**Comment by ematipico** [comment]
> This still is a case where key would not make sense, yet it wouldn't be declared inside a new Map(...).

> The only time a key should be needed is when an array actually gets rendered inside a component.

> That said, I don't know if it's possible to trace that, so detecting based on whether a tuple is declared inside a Map constructor may be the best option.

The original rule (eslint) checks if JSX elements are declared inside an array. As a consequence, they should be checked for `Map` and `Set` because, like arrays, they are iterable.

We aren't in a position to check if an iterable is rendered or not, so it's nothing or all. We can't check only arrays.

**Comment by dm-pse** [comment]
> The original rule (eslint) checks if JSX elements are declared inside an array. As a consequence, they should be checked for `Map` and `Set` because, like arrays, they are iterable.

But the level of iteration is one up from where the JSX element is declared. A Map is an iterable of key-value pairs. The key-value pairs would not themselves be iterated in any reasonable rendering scenario.

So, you might iterate `["None", <span>None</span>], ["Partial", <span>Partial</span>], ["Full", <span>Full</span>]`, but you would almost never iterate `["None", <span>None</span>]`, so `span` should not need a `key` prop.

> We aren't in a position to check if an iterable is rendered or not, so it's nothing or all. We can't check only arrays.

I agree that it would not be feasible to determine if something eventually gets rendered in some other part of the codebase, however, the suggestion in my "EDIT" was to specifically skip the `key` check for a two-item tuple defined inside a parent array. This would cover a Map instantiation without the key-value pair array needing to be inside a new Map(...) statement.

**Comment by Pranav2612000** [comment]
> however, the suggestion in my "EDIT" was to specifically skip the key check for a two-item tuple defined inside a parent array.

I guess this logic could have a lot of false positives. Do you think another check on top of this which checks that this is part of the Map/WeakMap etc constructor is better?


Another approach could be that we ignore this rule, if there is only one JSX element in the array ( since its sort of unique and doesn't need the `key` property ). Not sure if its trivial to implement

**Comment by devuxer** [comment]
> Another approach could be that we ignore this rule, if there is only one JSX element in the array ( since its sort of unique and doesn't need the `key` property ). Not sure if its trivial to implement

I like this approach. I considered an idea like this before but dismissed it because a string or number could still be rendered, but now that I think about it, if you have a two-item tuple where the first item is a non-JSX element and the second item is a JSX element, it would be a pretty safe bet that there does not need to be a key in this case.

**Comment by devuxer** [comment]
Was there ever any resolution on this?
